File: Scala/src/main/scala/TextbookTrack/Chapter10/BA10C.scala
package TextbookTrack.Chapter10

object BA10C {
  import scala.annotation.tailrec

  case class Label(labels: Vector[Char]) {
    private val labelIndices: Map[Char, Int] = labels.iterator.zipWithIndex.toMap
    val size: Int = labels.length

    def apply(label: Char): Int = labelIndices(label)

    def zipWithIndex: Iterator[(Char, Int)] = labels.iterator.zipWithIndex
  }

  case class ProbabilityMatrix(rowLabels: Label, columnLabels: Label, probabilities: Vector[Vector[Double]]) {
    val nrRows: Int = rowLabels.size
    val nrCols: Int = columnLabels.size

    def apply(row: Char, col: Char): Double = probabilities(rowLabels(row))(columnLabels(col))
  }

  case class HMM(alphabet: Label, states: Label, transition: ProbabilityMatrix, emission: ProbabilityMatrix)

  private def collectLabels(line: String): Vector[Char] =
    line.toVector.filter(_.isLetter)

  private def readMatrix(reader: Iterator[String], k: Int): Vector[Vector[Double]] = {
    val _: String = reader.next()
    reader
      .take(k)
      .map(_.split("\\s").drop(1).map(_.toDouble).toVector)
      .toVector
  }

  private def readData(reader: Iterator[String]): (String, HMM) = {
    val string: String = reader.next()
    val _: String = reader.next()
    val alphabet = Label(collectLabels(reader.next()))
    val _: String = reader.next()
    val states = Label(collectLabels(reader.next()))
    val _: String = reader.next()
    val transition = ProbabilityMatrix(states, states, readMatrix(reader, states.size))
    val _: String = reader.next()
    val emission = ProbabilityMatrix(states, alphabet, readMatrix(reader, states.size))
    val hmm = HMM(alphabet, states, transition, emission)
    (string, hmm)
  }

  private def reconstructOptimalHiddenPath(backtrack: Array[Array[Int]],
                                           nrCols: Int,
                                           maximalScoreIndex: Int,
                                           states: Label): String = {
    @tailrec
    def loop(hiddenPath: List[Char], jy: Int, index: Int): String = {
      if (jy < 0) hiddenPath.mkString
      else {
        val maximalIndex: Int = backtrack(index)(jy)
        loop(states.labels(maximalIndex) :: hiddenPath, jy - 1, maximalIndex)
      }
    }
    loop(List(states.labels(maximalScoreIndex)), nrCols - 1, maximalScoreIndex)
  }

  def calcMostLikelyHiddenPath(hmm: HMM, emittedString: String): String = {
    val HMM(_, states, transition, emission) = hmm
    val nrCols: Int = emittedString.length
    val logscoreMatrix: Array[Array[Double]] = Array.fill(states.size, nrCols)(0.0)
    val backtrack: Array[Array[Int]] = Array.fill(states.size, nrCols - 1)(0)
    val char: Char = emittedString(0)
    for {
      (state, ix) <- states.zipWithIndex
    } logscoreMatrix(ix)(0) = math.log(1.0 / states.size) + math.log(emission(state, char))

    for {
      (char, jy) <- emittedString.drop(1).zipWithIndex
      (state, ix) <- states.zipWithIndex
    } {
      val emissionLogprob: Double = math.log(emission(state, char))
      val logScores: Iterator[Double] =
        states
          .zipWithIndex
          .map{
            case (previousState, k) =>
              logscoreMatrix(k)(jy) + math.log(transition(previousState, state)) + emissionLogprob
          }
      val (maximumSumWeight, index): (Double, Int) = logScores.zipWithIndex.maxBy{ case (score, _) => score }
      logscoreMatrix(ix)(jy + 1) = maximumSumWeight
      backtrack(ix)(jy) = index
    }

    val (_, maximalScoreIndex): (Double, Int) =
      logscoreMatrix.map(_.last).zipWithIndex.maxBy{ case (score, _) => score }
    reconstructOptimalHiddenPath(backtrack, nrCols - 1, maximalScoreIndex, states)
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val (emittedString, hmm): (String, HMM) = readData(reader)
    val result: String = calcMostLikelyHiddenPath(hmm, emittedString)
    println(result)
  }
}

File: Python/textbook_track/chapter10/ba10c.py
# Implement the Viterbi Algorithm
import math
import sys

from collections import namedtuple

HMM = namedtuple('HMM', ['alphabet', 'states', 'transition', 'emission'])


class Label:
    def __init__(self, labels):
        self._labels = labels
        self._label_indices = {item: ix for ix, item in enumerate(labels)}
        self._size = len(labels)

    def __getitem__(self, label):
        return self._label_indices[label]

    def __repr__(self):
        return ', '.join(map(str, self._labels))

    def __iter__(self):
        yield from self._labels

    @property
    def labels(self):
        return self._labels

    @property
    def size(self):
        return self._size


class ProbabilityMatrix:
    def __init__(self, row_labels, column_labels, probabilities):
        self._row_labels = row_labels
        self._column_labels = column_labels
        self._probabilities = probabilities

    def __getitem__(self, label_pair):
        row_label, col_label = label_pair
        return self._probabilities[self._row_labels[row_label]][self._column_labels[col_label]]

    @property
    def nr_rows(self):
        return self._row_labels.size

    @property
    def nr_cols(self):
        return self._column_labels.size


def read_matrix(reader, k):
    _ = next(reader)
    matrix = []
    for _ in range(k):
        line = next(reader)
        probabilities = [float(item) for item in line.split()[1:]]
        matrix.append(probabilities)
    return matrix


def read_data(reader):
    string = next(reader).rstrip()
    _ = next(reader)
    alphabet = Label(next(reader).split())
    _ = next(reader)
    states = Label(next(reader).split())
    _ = next(reader)
    transition = ProbabilityMatrix(states, states, read_matrix(reader, states.size))
    _ = next(reader)
    emission = ProbabilityMatrix(states, alphabet, read_matrix(reader, states.size))
    hmm = HMM(alphabet=alphabet, states=states, transition=transition, emission=emission)
    return string, hmm


def calc_argmax(sequence):
    return max(enumerate(sequence), key=lambda x: x[1])


def reconstruct_optimal_hidden_path(backtrack, nr_cols, maximal_score_index, states):
    hidden_path = [states.labels[maximal_score_index]]
    jy = nr_cols
    while jy > 0:
        jy -= 1
        maximal_score_index = backtrack[maximal_score_index][jy]
        hidden_path.append(states.labels[maximal_score_index])
    return ''.join(hidden_path[::-1])


def calc_most_likely_hidden_path(hmm, emitted_string):
    _, states, transition, emission = hmm
    nr_cols = len(emitted_string)
    logscore_matrix = [[0.0] * nr_cols for _ in range(states.size)]
    backtrack = [[0] * (nr_cols - 1) for _ in range(states.size)]
    char = emitted_string[0]
    for ix, state in enumerate(states):
        logscore_matrix[ix][0] = math.log(1 / states.size) + math.log(emission[state, char])

    for jy, char in enumerate(emitted_string[1:], start=1):
        for ix, state in enumerate(states):
            emission_logprob = math.log(emission[state, char])
            logscores = [logscore_matrix[k][jy-1] + math.log(transition[previous_state, state]) + emission_logprob
                         for k, previous_state in enumerate(states)]
            index, maximum_sum_weight = calc_argmax(logscores)
            logscore_matrix[ix][jy] = maximum_sum_weight
            backtrack[ix][jy-1] = index
    maximal_score_index, _ = calc_argmax(map(lambda x: x[-1], logscore_matrix))
    return reconstruct_optimal_hidden_path(backtrack, nr_cols-1, maximal_score_index, states)


def main():
    reader = sys.stdin
    emitted_string, hmm = read_data(reader)
    result = calc_most_likely_hidden_path(hmm, emitted_string)
    print(result)


if __name__ == '__main__':
    main()
